File: solutions/python/two-pointers/find-the-index-of-the-first-occurrence-in-a-string.py
# LeetCode 28 — Find the Index of the First Occurrence in a String (Easy)
# Category: Two Pointers · Approach: Sliding Compare
# Time: O(n*m) | Space: O(1)
# Source: https://leetcode.com/problems/find-the-index-of-the-first-occurrence-in-a-string/

def str_str(haystack: str, needle: str) -> int:
    n, m = len(haystack), len(needle)
    for i in range(n - m + 1):
        if haystack[i:i + m] == needle:
            return i
    return -1

File: solutions/typescript/two-pointers/find-the-index-of-the-first-occurrence-in-a-string.ts
// LeetCode 28 — Find the Index of the First Occurrence in a String (Easy)
// Category: Two Pointers · Approach: Sliding Compare
// Time: O(n*m) | Space: O(1)
// Source: https://leetcode.com/problems/find-the-index-of-the-first-occurrence-in-a-string/

function strStr(haystack: string, needle: string): number {
  const n = haystack.length;
  const m = needle.length;
  for (let i = 0; i + m <= n; i++) {
    let j = 0;
    while (j < m && haystack[i + j] === needle[j]) j++;
    if (j === m) return i;
  }
  return -1;
}
